examples/CircleOptimisation.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2023 Siddharth Bhat. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddharth Bhat
+
+Riemannian Gradient Descent on S1.
+-/
+
+import SciLean
+open Lean
+open SciLean
+
+
+
+/-- choose a particular A -/
+-- the function (x, y) -> (x + y, x - y)
+def A (x : ℝ^{2}) : ℝ^{2} :=
+  ⊞ i, if i = 0 then x[0] + x[1] else x[0] - x[1]
+
+-- set_option trace.Meta.Tactic.fun_trans.rewrite true in
+function_properties A (x : ℝ^{2})
+argument x
+  IsLin, HasAdjoint, IsSmooth, HasAdjDiff,
+  abbrev ∂ := λ dx => A dx by unfold A; fun_trans,
+  def † :=
+    λ x' => ⊞ i,
+      if i = 0 then x'[0] - x'[1] else x'[0] + x'[1] by sorry,
+  abbrev ∂† by unfold adjointDifferential; fun_trans; fun_trans; simp
+
+
+set_option trace.Meta.Tactic.fun_trans.rewrite true in
+def loss (x : ℝ^{2}) : ℝ := ⟪x, A x⟫
+
+function_properties loss (x : ℝ^{2})
+argument x
+  IsSmooth, HasAdjDiff,
+  abbrev ∂ := λ dx => (2 : ℝ) * ⟪dx, A x⟫ by {
+    simp[loss];
+    fun_trans; -- 'trans' makes me think of 'transitive'.
+    simp;
+    funext dx;
+    -- ⊢ ⟪dx, A x⟫ + ⟪x, A dx⟫ = 2 * ⟪dx, A x⟫
+    -- how do I prove this? The natural proof for me is to do this componentwise
+    sorry
+  }
+
+noncomputable def grad_loss (x : ℝ^{2}) : ℝ^{2} := ∇ loss x
```

Comment:
You can add a property to `loss`
```
  abbrev ∂† by unfold gradient; unfold loss; fun_trans; simp
```

Then you can define `grad_loss` as
```
def grad_loss (x : ℝ^{2}) : ℝ^{2} := (∇ loss x) rewrite_by unfold gradient; unfold loss; fun_trans
```
and it is computable now.

Reply:
Thanks! I am unsure why there are three things I can define: `∂`, `†`, and ` ∂†`. My impression is that if all I want is reverse mode AD, I can get by with just ` ∂†`?

Reply:
Yes that is correct but for proper reverse mode AD you actually want ℛ
```
ℛ f = (f x, λ dy => ∂† f x dy)
```